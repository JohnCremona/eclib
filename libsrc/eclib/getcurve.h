// getcurve.h: declaration of function getcurve() for curve input
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#if     !defined(_ECLIB_GETCURVE_H)
#define _ECLIB_GETCURVE_H      1       //flags that this file has been included

#include <eclib/curve.h>

// Read in a curve as [a1,a2,a3,a4,a6] with ai integers:
int getcurve(Curvedata& CD, int verb);

// Read in a curve as [a1,a2,a3,a4,a6] with ai rational:
int getcurve(vector<bigrational>& ai, int verb);

#endif
