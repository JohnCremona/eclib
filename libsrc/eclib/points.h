// points.h:  declarations of Point class for points on elliptic curves
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// originally adapted from Elliptic.h by Oisin McGuiness

// allow for multiple includes
#ifndef _ECLIB_POINTS_
#define _ECLIB_POINTS_

#include "polys.h"
#include "p2points.h"
#include "divpol.h"

//
// class for  points on elliptic curves
//

class Point;
Point transform(const Point& p,  Curvedata* newc, 
			 const bigint& u, 
			 const bigint& r, const bigint& s, const bigint& t, 
			 int back=0); 

class Point : public P2Point { 
  Curvedata *E;    // pointer to the curve that the point is on
  int ord;         // order: 0 if not calculated yet, -1 if infinite
  bigfloat height; // -1.0 if not calculated yet, 0.0 for torsion point
public:
  // constructors
  Point(void)
    : P2Point(), E(0), ord(0), height(to_bigfloat(-1.0))
    { ; }
  explicit Point(Curvedata &EE)      // set to point at infinity
    : P2Point(0,1,0), E(&EE), ord(1), height(to_bigfloat(0.0))
    { ; }
  explicit Point(Curvedata *EE)      // set to point at infinity
    : P2Point(0,1,0), E(EE), ord(1), height(to_bigfloat(0.0))
    { ; }
  Point(Curvedata &EE, const bigint& x, const bigint& y, const bigint& z)
    : P2Point(x,y,z), E(&EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(Curvedata &EE, const P2Point& P)
    : P2Point(P), E(&EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(Curvedata *EE, const bigint& x, const bigint& y, const bigint& z)
    : P2Point(x,y,z), E(EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(Curvedata *EE, const P2Point& p)
    : P2Point(p), E(EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(Curvedata &EE, const bigint& x, const bigint& y)
    : P2Point(x,y), E(&EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(Curvedata *EE, const bigint& x, const bigint& y)
    : P2Point(x,y), E(EE), ord(0), height(to_bigfloat(-1.0))
    { ; }
  Point(const Point& Q)
    : P2Point(Q), E(Q.E), ord(Q.ord), height(Q.height)
    { ; }
  ~Point(void) {;}

  // input and output are inherited from P2Point class but the input
  // function must initialize the ord and height fields too
  friend istream& operator>>(istream & is, Point& P)
  {
    is>>(P2Point&)P;
    P.ord=0;
    P.height=to_bigfloat(-1.0);
    // NB P's Curve should have been set when it was constructed.
    return is;
  }

  // test of equality of points        
  int operator==(const Point& Q) const
  {
    if(E != Q.E) return 0 ;      // different curves!
    return eq(*this,Q);
  }
  int operator!=(const Point& Q) const { return !(*this == Q); }

  // assignment (p.init(.) is quicker than p=Point(.) for existing p)
  void init(Curvedata &EE,
            const bigint& x, const bigint& y, const bigint& z)
    {E=&EE; X=x; Y=y; Z=z; reduce(); ord=0; height=-1.0; }
  void init(Curvedata *EE,
            const bigint& x, const bigint& y, const bigint& z)
    {E=EE; X=x; Y=y; Z=z; reduce(); ord=0; height=-1.0; }
  void init(Curvedata &EE,
            const bigint& x, const bigint& y)
    {E=&EE; X=x; Y=y; Z = 1; ord=0; height=-1.0; }
  void init(Curvedata *EE,
            const bigint& x, const bigint& y)
    {E=EE; X=x; Y=y; Z = 1; ord=0; height=-1.0; }
  void operator=(const Point& Q) // P1 = P2
    { E=Q.E; X=Q.X ; Y=Q.Y; Z=Q.Z; ord=Q.ord; height=Q.height; }

  friend Point transform(const Point& p,  Curvedata* newc, 
			 const bigint& u, 
			 const bigint& r, const bigint& s, const bigint& t, 
			 int back); 

  void operator+=(const Point&) ; // P1 += P2 ; order and height unknown
  void operator-=(const Point&) ; // P1 -= P2 ; ditto

  // addition of points, etc
  Point operator+(const Point &) const ; // P1 + P2
  Point operator-(const Point &) const ; // P1 - P2
  Point operator-(void) const ; // -P
  Point twice(void) const ; // doubles P
  friend Point operator*(int, const Point&) ; // n*P

  // access functions
  Curve getcurve() const {return *E;}
  friend int order(Point& p);       // calculate and set if not set
  friend int order(Point& p,  vector<Point>&multiples);
                        // also create and return list of multiples
  friend bigfloat height(Point& P);    //calculate and set if not set
  friend bigfloat realheight(const Point& P);
  friend bigfloat pheight(const Point& P, const bigint& p);

  vector<Point> division_points(int m); // list of Q s.t. n*Q=this

  // useful logical tests
  int is_zero() const { return isinfinite(); }
  int isvalid() const ; // P on its curve ?
  int is_torsion() { return order(*this)>0; } // will compute and set the order if needed
  int is_on_real_identity_component() const;
  int is_on_egg() const {return !is_on_real_identity_component();}
// return 1 if P mod p is nonsingular (or for p=0 if it is on the real identity component):
  int has_good_reduction(long p) const;
  int has_good_reduction(const bigint& p) const;
  // return 1 if P has good reduction at all p in list, if not then p0 holds first bad prime
  int has_good_reduction(const vector<bigint>& plist, bigint& p0, int check_real=0) const;

}; // end of point class

// list of 0,1 or 2 points with given x-coordinate:
vector<Point> points_from_x(Curvedata &E, const bigrational& x);

// the real x and y coords of the point
void realify_point(const Point& P, bigfloat&x, bigfloat& y);

// the real component of the canonical height
bigfloat realheight(const bigfloat& x, const Curvedata* E);


// the height pairing of two points
bigfloat height_pairing(Point& P, Point& Q);

// regulator of a list of n points
bigfloat regulator(vector<Point>& points);  // not a const array; heights get set.

// torsion functions returning
// (exact==0) list of points in E[m] for m=2, m=3 and general m
// (exact==1) list of points of exact order m for m=2, m=3 and general m

// N.B. Don't make the params const here

vector<Point> two_torsion(Curvedata& E, int exact=0);
vector<bigint> three_torsion_x(Curvedata& E);
vector<Point> three_torsion(Curvedata& E, int exact=0);

// List m-torsion points, i.e. points in E[m], or points of exact order m if exact==1
vector<Point> m_torsion(Curvedata& E, long m, int exact=0);
vector<Point> torsion_points(Curvedata& E);

inline long ntorsion(Curvedata& E)
{
  return E.get_ntorsion();
}

// end of file: points.h

#endif
