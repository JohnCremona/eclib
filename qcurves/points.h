// points.h:  declarations of Point class for points on elliptic curves
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
// originally adapted from Elliptic.h by Oisin McGuiness

// allow for multiple includes
#ifndef _PELLIPTIC_
#define _PELLIPTIC_

#include "curve.h"

//
// class for  points on elliptic curves
//

class Point{ 
  bigint X ; // homogeneous coordinates
  bigint Y ;
  bigint Z ;
  Curvedata *E; // pointer to the curve that the point is on
                   // warning: no check is made that client doesn't alter curve
  int ord; // 0 if not calculated yet, -1 if infinite
  bigfloat height; // -1.0 if not calculated yet, 0.0 for torsion point
  void reduce(); //divide out coordinate gcd
public:
  // constructors 
  Point(void)       // not a real point
    : E(0), ord(0), height(to_bigfloat(-1.0))
    { X=0; Y=0; Z=0;} 
  Point(Curvedata &EE)      // set to point at infinity
    : E(&EE), ord(1), height(to_bigfloat(0.0))
    { X=0; Y=1; Z=0;}      
  Point(Curvedata *EE)      // set to point at infinity
    : E(EE), ord(1), height(to_bigfloat(0.0))
    { X=0; Y=1; Z=0;}      
  Point(Curvedata &EE, const bigint& x, const bigint& y, const bigint& z)
    : X(x), Y(y), Z(z), E(&EE), ord(0), height(to_bigfloat(-1.0))
    { reduce(); }
  Point(Curvedata *EE, const bigint& x, const bigint& y, const bigint& z)
    : X(x), Y(y), Z(z), E(EE), ord(0), height(to_bigfloat(-1.0))
    { reduce(); }
  Point(Curvedata &EE, const bigint& x, const bigint& y)
    : X(x), Y(y), E(&EE), ord(0), height(to_bigfloat(-1.0))
    { Z=1;} // no need to reduce already integral
  Point(Curvedata *EE, const bigint& x, const bigint& y)
    : X(x), Y(y), E(EE), ord(0), height(to_bigfloat(-1.0))
    { Z=1;} // no need to reduce already integral
  Point(const Point& Q)
    :X(Q.X), Y(Q.Y), Z(Q.Z), E(Q.E), ord(Q.ord), height(Q.height)
    { ; }
  ~Point(void) {;}
                
  // input and output
  friend inline ostream& operator<<(ostream & os, const Point& P)
    {return os << "[" << P.X << ":" << P.Y << ":" << P.Z << "]" ;}

  friend inline void output_pari(ostream&os, const Point& P)
    {
      bigint xp=P.X, yp=P.Y, zp=P.Z;
      if(is_zero(zp)) {os<<"[0]"; return;}
      if(is_one(zp)) {os<<"["<<xp<<","<<yp<<"]"; return;}
      bigint z=gcd(xp,zp);
      os<<"["<<(xp/z)<<"/"<<(zp/z)<<","<<yp<<"/"<<zp<<"]";
    }

// Point input: 3 formats allowed are 
// [x:y:z], [x,y], [x/z,y/z] with any stype of brackets
  friend istream& operator>>(istream & is, Point& P);

  // test of equality of points        
  int operator==(const Point& Q) const
  {
    if(E != Q.E) return 0 ;      // different curves!
    return eq(*this,Q);
  }
  int operator!=(const Point& Q) const { return !(*this == Q); }
  friend int eq(const Point&P, const Point&Q);

  // assignment (p.init(.) is quicker than p=Point(.) for existing p)
  void init(Curvedata &EE,
            const bigint& x, const bigint& y, const bigint& z)
    {E=&EE; X=x; Y=y; Z=z; reduce(); ord=0; height=-1.0; }
  void init(Curvedata *EE,
            const bigint& x, const bigint& y, const bigint& z)
    {E=EE; X=x; Y=y; Z=z; reduce(); ord=0; height=-1.0; }
  void init(Curvedata &EE,
            const bigint& x, const bigint& y)
    {E=&EE; X=x; Y=y; Z = 1; ord=0; height=-1.0; }
  void init(Curvedata *EE,
            const bigint& x, const bigint& y)
    {E=EE; X=x; Y=y; Z = 1; ord=0; height=-1.0; }
  void operator=(const Point& Q) // P1 = P2
    { E=Q.E; X=Q.X ; Y=Q.Y; Z=Q.Z; ord=Q.ord; height=Q.height; }

  friend Point shift(const Point& p,  Curvedata* newc, 
		     const bigint& u, const bigint& r, 
		     const bigint& s, const bigint& t, int back=0); 

  void operator+=(const Point&) ; // P1 += P2 ; order and height unknown
  void operator-=(const Point&) ; // P1 -= P2 ; ditto

  // addition of points, etc
  Point operator+(const Point &) const ; // P1 + P2
  Point operator-(const Point &) const ; // P1 - P2
  Point operator-(void) const ; // -P
  Point twice(void) const ; // doubles P
  friend Point operator*(int, const Point&) ; // n*P
                
  // access functions
  Curve getcurve() const {return *E;}
  void getcoordinates(bigint& x, bigint& y, bigint& z) const
    {x=X; y=Y; z=Z; }
  friend inline bigint getX(const Point& p) {return p.X; }
  friend inline bigint getY(const Point& p) {return p.Y; }
  friend inline bigint getZ(const Point& p) {return p.Z; }
  friend int order(Point& p);       // calculate and set if not set
  friend int order(Point& p,  vector<Point>&multiples);
                        // also create and return list of multiples
  friend bigfloat height(Point& P);    //calculate and set if not set
  friend bigfloat realheight(const Point& P);
  friend bigfloat pheight(const Point& P, const bigint& p);

  // useful logical tests
  int iszero() const { return Z==0; }
  int isintegral() const { return Z==1; }
  int isvalid() const ; // P on its curve ?

}; // end of point class

// the real x and y coords of the point
void realify_point(const Point& P, bigfloat&x, bigfloat& y);

// the real component of the canonical height
bigfloat realheight(const bigfloat& x, const Curvedata* E);


// the height pairing of two points
bigfloat height_pairing(Point& P, Point& Q);  

// regulator of a list of n points
bigfloat regulator(vector<Point>& points);  // not a const array; heights get set.

void raw_shift_point(const bigint& x, const bigint& y, const bigint& z, 
                const bigint& u, const bigint& r, 
                const bigint& s, const bigint& t, 
                bigint& newx, bigint& newy, bigint& newz, int back=0);


// torsion functions
// N.B. Don't make the params const here

vector<Point> two_torsion(Curvedata& E);
vector<bigint> three_torsion_x(Curvedata& E);
vector<Point> three_torsion(Curvedata& E);
vector<Point> torsion_points(Curvedata& E);

inline long ntorsion(Curvedata& E)
{
  return E.get_ntorsion();
}

// end of file: points.h

#endif
