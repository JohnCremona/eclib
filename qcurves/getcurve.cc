// getcurve.cc: implementation of function getcurve() for curve input
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "curve.h"
#include "getcurve.h"

int getcurve(Curvedata& CD, int verb)
{
  Curve C0;
  if(verb) cout  << "Enter curve: ";
  cin>>ws;  if(cin.eof()) return 0; // quit if EOF reached
  cin >> C0;
  if (verb) cout << endl;
  if(C0.isnull()) return 0;  // quit if null curve entered
  CD = Curvedata(C0,0);      // DON'T change coords
  if(CD.isnull()) // input curve was singular, non-null
    {
      cout<<C0<<" is singular"<<endl;
      return 0;
    }
  return 1;
}

//#define DEBUG_Q_INPUT

int getcurve(vector<bigrational>& ai, int verb)
{
  // read the coefficients, either as "a1 a2 a3 a4 a6" or as
  // "[a1,a2,a3,a4,a6]" using code essentially the same as in
  // ../qcurves/curve.cc
  ai.resize(5);
  if(verb) cerr  << "Enter curve: ";
  cin>>ws;  if(cin.eof()) return 0; // quit if EOF reached
  char c;
  cin.get(c);
#ifdef DEBUG_Q_INPUT
  cout<<"First char read = "<<c<<"\n";
#endif
  if(c=='[')
    {
#ifdef DEBUG_Q_INPUT
      cout<<"Reading [a1,a2,a3,a4,a6]...\n";
#endif
      cin >> ai[0] >> c;
      if(c!=',')
	{
	  cout << "syntax error on curve input" << endl;
	  abort();
	}
      cin >> ai[1] >> c; 
      if(c!=',')
	{
	  cout << "syntax error on curve input" << endl;
	  abort();
	}
      cin >> ai[2] >> c; 
      if(c!=',')
	{
	  cout << "syntax error on curve input" << endl;
	  abort();
	}
      cin >> ai[3] >> c; 
      if(c!=',')
	{
	  cout << "syntax error on curve input" << endl;
	  abort();
	}
      cin >> ai[4] >> c; 
      if(c!=']')
	{
	  cout << "syntax error on curve input" << endl;
	  abort();
	}
#ifdef DEBUG_Q_INPUT
      cout<<ai<<endl;
#endif
    }
  else
    {
#ifdef DEBUG_Q_INPUT
      cout<<"Reading a1 a2 a3 a4 a6 ...\n";
#endif
      cin.unget();
      cin >> ai[0] >> ai[1] >> ai[2] >> ai[3] >> ai[4];
#ifdef DEBUG_Q_INPUT
      cout<<ai<<endl;
#endif
    }
  // test for null curve input
  if((num(ai[0])==0)&&(num(ai[1])==0)&&(num(ai[2])==0)&&(num(ai[3])==0)&&(num(ai[4])==0))  return 0;  // quit if null curve entered
  return 1;
}
