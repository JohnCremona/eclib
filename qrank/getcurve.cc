// getcurve.cc: implementation of function getcurve() for curve input
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include "curve.h"
#include "getcurve.h"

int getcurve(Curvedata& CD, int verb)
{
  Curve C0;
  if(verb) cout  << "Enter curve: ";
  cin>>ws;  if(cin.eof()) return 0; // quit if EOF reached
  cin >> C0;
  if (verb) cout << endl;
  if(C0.isnull()) return 0;  // quit if null curve entered
  cout << "Curve "<< C0<<" :\t";
  CD = Curvedata(C0,0);      // DON'T change coords
  if(CD.isnull()) // input curve was singular, non-null
    {
      cout<<" singular"<<endl;
      return 0;
    }
  return 1;
}
