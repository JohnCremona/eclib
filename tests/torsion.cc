// torsion.cc: program to find & display torsion points
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <eclib/points.h>
#include <eclib/getcurve.h>


int main(){
  cerr<<"Program to find and/or count torsion on a curve.\n\n";
#ifdef MPFP
  set_precision("Enter precision in bits");
#endif
  cerr<<"Enter 0 for short output (#torsion only)\n";
  cerr<<"   or 1 for long  output (list of the torsion points): ";
  int showpoints; cin >> showpoints;
  initprimes("PRIMES",0);
  vector<bigrational> ai(5);
  bigint u, r, s, t, v;
  int verb=1;

  while (getcurve(ai,verb))
    {
      Curvedata E0(ai,v);
      Curvedata E1 = E0.minimalize(u,r,s,t);
      cout<<"Curve ["<<ai[0]<<","<<ai[1]<<","<<ai[2]<<","<<ai[3]<<","<<ai[4]<<"]"<<flush;
      vector<Point> tor = torsion_points(E1); 
      int ntor = tor.size();
      cout<<" \t has " << ntor << " torsion point(s)\n";
      if(showpoints)
	{
          int n2 = 0; // counts 2-torsion
	  for(int i=0; i<ntor; i++)
	    {
	      Point P = transform(tor[i],&E0,u,r,s,t,1);
	      cout << scale(P,v,0) << "\t";
	      if(!P.isvalid()) cout << " --warning: NOT on curve!\n";
	      else
		{
		  int ord = order(P);
		  if(ord==2) n2++;
		  cout << "(order " << ord << ")\n";
		}
	    }
	  if(n2>2) cout<<"Non-cyclic: C2 x C"<<(ntor/2)<<"\n"; 
	  else cout << "Cyclic: C"<<ntor<<"\n";
	}
    }
} //ends main

	


