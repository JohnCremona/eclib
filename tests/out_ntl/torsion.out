Curve [0,0,0,1,1] 	 has 1 torsion point(s)
[0:1:0]	(order 1)
Cyclic: C1
Curve [0,1,1,101470,57781877] 	 has 1 torsion point(s)
[0:1:0]	(order 1)
Cyclic: C1
Curve [0,0,0,1,0] 	 has 2 torsion point(s)
[0:1:0]	(order 1)
[0:0:1]	(order 2)
Cyclic: C2
Curve [0,0,0,0,100] 	 has 3 torsion point(s)
[0:1:0]	(order 1)
[0:10:1]	(order 3)
[0:-10:1]	(order 3)
Cyclic: C3
Curve [0,0,0,1,2] 	 has 4 torsion point(s)
[0:1:0]	(order 1)
[1:-2:1]	(order 4)
[-1:0:1]	(order 2)
[1:2:1]	(order 4)
Cyclic: C4
Curve [0,0,0,-1,0] 	 has 4 torsion point(s)
[0:1:0]	(order 1)
[-1:0:1]	(order 2)
[0:0:1]	(order 2)
[1:0:1]	(order 2)
Non-cyclic: C2 x C2
Curve [0,-1,1,0,0] 	 has 5 torsion point(s)
[0:1:0]	(order 1)
[1:-1:1]	(order 5)
[0:-1:1]	(order 5)
[0:0:1]	(order 5)
[1:0:1]	(order 5)
Cyclic: C5
Curve [0,0,0,0,1] 	 has 6 torsion point(s)
[0:1:0]	(order 1)
[2:-3:1]	(order 6)
[0:-1:1]	(order 3)
[-1:0:1]	(order 2)
[0:1:1]	(order 3)
[2:3:1]	(order 6)
Cyclic: C6
Curve [1,-1,1,-3,3] 	 has 7 torsion point(s)
[0:1:0]	(order 1)
[3:-6:1]	(order 7)
[1:-2:1]	(order 7)
[-1:-2:1]	(order 7)
[-1:2:1]	(order 7)
[1:0:1]	(order 7)
[3:2:1]	(order 7)
Cyclic: C7
Curve [0,1,0,-14624,669300] 	 has 8 torsion point(s)
[0:1:0]	(order 1)
[418:-8232:1]	(order 8)
[124:-882:1]	(order 4)
[82:-168:1]	(order 8)
[75:0:1]	(order 2)
[82:168:1]	(order 8)
[124:882:1]	(order 4)
[418:8232:1]	(order 8)
Cyclic: C8
Curve [0,1,0,-160,308] 	 has 8 torsion point(s)
[0:1:0]	(order 1)
[26:-120:1]	(order 4)
[11:0:1]	(order 2)
[26:120:1]	(order 4)
[-14:0:1]	(order 2)
[-4:30:1]	(order 4)
[2:0:1]	(order 2)
[-4:-30:1]	(order 4)
Non-cyclic: C2 x C4
Curve [1,-1,1,-14,29] 	 has 9 torsion point(s)
[0:1:0]	(order 1)
[9:-29:1]	(order 9)
[3:-5:1]	(order 9)
[1:-5:1]	(order 3)
[-3:-5:1]	(order 9)
[-3:7:1]	(order 9)
[1:3:1]	(order 3)
[3:1:1]	(order 9)
[9:19:1]	(order 9)
Cyclic: C9
Curve [1,0,0,115,561] 	 has 10 torsion point(s)
[0:1:0]	(order 1)
[70:-629:1]	(order 10)
[16:-89:1]	(order 5)
[4:-35:1]	(order 10)
[-2:-17:1]	(order 5)
[-34:17:8]	(order 2)
[-2:19:1]	(order 5)
[4:31:1]	(order 10)
[16:73:1]	(order 5)
[70:559:1]	(order 10)
Cyclic: C10
Curve [1,0,1,-19,26] 	 has 12 torsion point(s)
[0:1:0]	(order 1)
[13:-52:1]	(order 6)
[4:-7:1]	(order 3)
[3:-2:1]	(order 2)
[4:2:1]	(order 3)
[13:38:1]	(order 6)
[-5:2:1]	(order 2)
[-2:8:1]	(order 6)
[1:2:1]	(order 6)
[14:-11:8]	(order 2)
[1:-4:1]	(order 6)
[-2:-7:1]	(order 6)
Non-cyclic: C2 x C6
Curve [1,0,0,-1070,7812] 	 has 16 torsion point(s)
[0:1:0]	(order 1)
[244:-3902:1]	(order 8)
[64:-482:1]	(order 4)
[34:-122:1]	(order 8)
[28:-14:1]	(order 2)
[34:88:1]	(order 8)
[64:418:1]	(order 4)
[244:3658:1]	(order 8)
[-36:18:1]	(order 2)
[-26:148:1]	(order 8)
[-8:130:1]	(order 4)
[4:58:1]	(order 8)
[62:-31:8]	(order 2)
[4:-62:1]	(order 8)
[-8:-122:1]	(order 4)
[-26:-122:1]	(order 8)
Non-cyclic: C2 x C8
Curve [1,0,0,-372368141774940800,87459461608665181808640000] 	 has 16 torsion point(s)
[0:1:0]	(order 1)
[654296320:-11132578294400:1]	(order 8)
[369088000:-549923580800:1]	(order 4)
[353359600:-34032060800:1]	(order 8)
[352441600:-176220800:1]	(order 2)
[353359600:33678701200:1]	(order 8)
[369088000:549554492800:1]	(order 4)
[654296320:11131923998080:1]	(order 8)
[-5636968450:2818484225:8]	(order 2)
[117433600:6734213027200:1]	(order 8)
[335795200:532525225600:1]	(order 4)
[351262480:33612641920:1]	(order 8)
[352179456:-176089728:1]	(order 2)
[351262480:-33963904400:1]	(order 8)
[335795200:-532861020800:1]	(order 4)
[117433600:-6734330460800:1]	(order 8)
Non-cyclic: C2 x C8
Curve [0,0,0,-4/9,0] 	 has 4 torsion point(s)
[0:1:0]	(order 1)
[-2:0:3]	(order 2)
[0:0:1]	(order 2)
[2:0:3]	(order 2)
Non-cyclic: C2 x C2
